File: NoOverlap/src/NoOverlap.java
// Overlapping Algorithm from The Coding Train - https://youtu.be/XATr_jdh-44
// Press 'a' to clear the circles
// Press 's' to change the color scheme

import java.util.ArrayList;

import processing.core.PApplet;

import color.ColorGenerator;
import color.ColorGeneratorFactory;

public class NoOverlap extends PApplet {
    private ArrayList<Circle> circles;
    private ColorGenerator colorGenerator;
    private ColorGeneratorFactory colorGeneratorFactory;

    public static void main(String[] args) {
        String[] processingArgs = {"NoOverlap"};
        PApplet.main(processingArgs);
    }

    public void settings() {
        size(800, 800);
    }

    public void setup() {
        circles = new ArrayList<>();
        colorGeneratorFactory = new ColorGeneratorFactory(this);
        colorGenerator = colorGeneratorFactory.getRandomColorGenerator();
    }

    public void draw() {
        background(0);
        addCircle();
        displayCircles();
    }

    public void keyPressed() {

        if (key == 'a') {
            circles.clear();
        } else if (key == 's') {
            changeColorScheme();
        }
    }

    private void addCircle() {
        Circle potentialCircle = new Circle(this);
        boolean isOverlapping = false;

        for (Circle circle: circles) {

            if (potentialCircle.overlap(circle)) {
                isOverlapping = true;
                break;
            }
        }

        if (!isOverlapping) {
            potentialCircle.setColor(colorGenerator.getRandomColor());
            circles.add(potentialCircle);
        }
    }

    private void displayCircles() {

        for (Circle circle: circles) {
            circle.display();
        }
    }

    private void changeColorScheme() {
        colorGenerator = colorGeneratorFactory.getRandomColorGenerator();

        for (Circle c: circles) {
            int color = colorGenerator.getRandomColor();
            c.setColor(color);
        }
    }
}

File: docs/NoOverlap/no_overlap.js
// No Overlap Sketch
// Press 'a' to clear the circles
// Press 's' to change the color scheme

var circles;
var colorGeneratorFactory;
var colorGenerator;

function setup() {
    createCanvas(windowWidth-20, windowHeight-20);
    frameRate(60);
    circles = [];
    colorGeneratorFactory = new ColorGeneratorFactory();
    colorGenerator = colorGeneratorFactory.getRandomColorGenerator();
}

function draw() {
    background(0);
    addCircle();
    displayCircles();
}

function keyTyped() {

    if (key == 'a') {
        circles.splice(0, circles.length);
    } else if (key == 's') {
        changeColorScheme();
    }
}

function addCircle() {
    let potentialCircle = new Circle();
    let isOverlapping = false;

    circles.forEach((circle) => {

        if (potentialCircle.isOverlapping(circle)) {
            isOverlapping = true;
        }
    });

    if (!isOverlapping) {
        let color = colorGenerator.randomColor();
        potentialCircle.setColor(color);
        circles.push(potentialCircle);
    }
}

function displayCircles() {

    circles.forEach((circle) => {
        circle.display();
    });
}

function changeColorScheme() {
    colorGenerator = colorGeneratorFactory.getRandomColorGenerator();

    circles.forEach((circle) => {
        let color = colorGenerator.randomColor();
        circle.setColor(color);
    });
}
